internal/graph/memoryWithDelete.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package graph
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/descriptor"
+	"oras.land/oras-go/v2/internal/status"
+	"oras.land/oras-go/v2/internal/syncutil"
+)
+
+// MemoryWithDelete is a MemoryWithDelete based PredecessorFinder.
+type MemoryWithDelete struct {
+	indexed      sync.Map // map[descriptor.Descriptor]any, this variable is only used by IndexAll
+	predecessors map[descriptor.Descriptor]map[descriptor.Descriptor]ocispec.Descriptor
+	successors   map[descriptor.Descriptor]map[descriptor.Descriptor]ocispec.Descriptor
```

Comment:
Can this be of type `map[descriptor.Descriptor][]ocispec.Descriptor`?
See https://github.com/oras-project/oras-go/pull/582#discussion_r1316895910